components/browser/engine-gecko-nightly/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngineSession.kt

```diff
@@ -347,6 +359,66 @@ class GeckoEngineSession(
         }
     }
 
+    @Suppress("ComplexMethod")
+    internal fun createHistoryDelegate() = object : GeckoSession.HistoryDelegate {
+        @Suppress("TooGenericExceptionCaught")
+        override fun onVisited(
+            session: GeckoSession,
+            url: String,
+            lastVisitedURL: String?,
+            flags: Int
+        ): GeckoResult<Boolean>? {
+            if ((flags and GeckoSession.HistoryDelegate.VISIT_TOP_LEVEL) == 0 ||
+                (flags and GeckoSession.HistoryDelegate.VISIT_UNRECOVERABLE_ERROR) != 0) {
+
+                return GeckoResult.fromValue(false)
+            }
+
+            val isReload = lastVisitedURL?.let { it == url } ?: false
+            val result = GeckoResult<Boolean>()
+            scope.launch {
+                var exception: Throwable? = null
+                try {
+                    settings.historyTrackingDelegate?.onVisited(url, isReload, privateMode)
+                } catch (t: Throwable) {
+                    exception = t
+                }
+
+                if (exception == null) {
+                    // Don't mark links as visited in private mode.
+                    val markAsVisited = !privateMode
+                    result.complete(markAsVisited)
+                } else {
+                    result.completeExceptionally(exception)
+                }
+            }
+            return result
+        }
+
+        @Suppress("TooGenericExceptionCaught")
+        override fun getVisited(
+            session: GeckoSession,
+            urls: Array<out String>
+        ): GeckoResult<BooleanArray>? {
+            val result = GeckoResult<BooleanArray>()
+            scope.launch {
+                var exception: Throwable? = null
+                val visits: List<Boolean>? = try {
+                    settings.historyTrackingDelegate?.getVisited(urls.toList(), privateMode)?.await()
+                } catch (t: Throwable) {
```

Comment:
Generally you watch to catch as narrow set of exceptions as possible. `Throwable` covers unchecked exceptions (subclasses of `RuntimeException` and `Error`, such as `NullPointerException`, `ThreadDeath`, etc), which is not what we want here. If there's nothing better, use `Exception` instead.

Reply:
Hmm, I'm not sure what all we'd like to catch here; IIUC, `getVisited` can throw anything. On the one hand, I don't think we want to crash if `getVisited` throws; on the other, I don't see other handlers doing the same, and, given that Gecko's [`OnError` handlers do nothing](https://searchfox.org/mozilla-central/rev/0859e6b10fb901875c80de8f8fc33cbb77b2505e/mobile/android/components/geckoview/GeckoViewHistory.cpp#363-367,565-569), maybe it's best to remove this entirely.